editor/src/components/navigator/navigator-utils.ts

```diff
@@ -79,317 +106,677 @@ export function navigatorDepth(
-        if (
-          elementMetadata != null &&
-          isRight(elementMetadata.element) &&
-          isJSXElement(elementMetadata.element.value) &&
-          elementMetadata.element.value.children.length > 0
-        ) {
-          const jsxElement = elementMetadata.element.value
-          const children = jsxElement.children
-          children.forEach((child) => {
-            const childPath = EP.appendToPath(path, child.uid)
-            const dataRefEntry = dataReferenceNavigatorEntry(
-              childPath,
-              renderedAtChildNode(path, child.uid),
-              path,
-              child,
-            )
-            addNavigatorTargetsUnlessCollapsed(dataRefEntry)
-          })
+      }
+      case 'conditional-entry': {
+        return {
+          type: 'conditional-entry',
+          navigatorEntry: entry.navigatorEntry,
+          trueCase: entry.trueCase.map(walkSubtreeMaybeCondense),
+          falseCase: entry.falseCase.map(walkSubtreeMaybeCondense),
+          elementHidden: entry.elementHidden,
+          subtreeHidden: entry.subtreeHidden,
         }
-      } else {
-        const notRenderPropChildren = Object.values(subTree.children).filter(
-          (c) => !processedPathsAsRenderProp.has(c.pathString),
-        )
-        if (
-          notRenderPropChildren.length > 0 &&
-          renderPropFound // only show a dedicated label for the children prop if the component has render props too
-        ) {
-          const entry = renderPropNavigatorEntry(
-            EP.appendToPath(path, renderPropId('children')),
-            'children',
-            notRenderPropChildren[0].path, // pick the first element path
-          )
-          addNavigatorTargetsUnlessCollapsed(entry)
+      }
+      case 'condensed-trunk': {
+        return {
+          type: 'condensed-trunk',
+          navigatorEntry: entry.navigatorEntry,
+          child: walkSubtreeMaybeCondense(entry.child),
         }
-        fastForEach(notRenderPropChildren, (child) => {
-          walkAndAddKeys(child, newCollapsedAncestor, null)
-        })
       }
+      case 'condensed-leaf':
+        return entry
+      default:
+        assertNever(entry)
     }
   }
 
-  const canvasRoots = getCanvasRoots(projectTree)
-  fastForEach(canvasRoots, (childElement) => {
-    const subTree = getSubTree(projectTree, childElement.path)
+  return navigatorTree.map(walkSubtreeMaybeCondense)
+}
 
-    walkAndAddKeys(subTree, false, null)
-  })
+function flattenCondensedTrunk(entry: CondensedTrunkNavigatorTree): {
+  singleRow: Array<NavigatorEntry>
+  child: NavigatorTree
+} {
+  const singleRow = [entry.navigatorEntry]
+  let currentChild: NavigatorTree = entry.child
+  while (currentChild.type === 'condensed-trunk') {
+    singleRow.push(currentChild.navigatorEntry)
+    currentChild = currentChild.child
+  }
+  return { singleRow: singleRow, child: currentChild }
+}
+
+function getNavigatorRowsForTree(
+  navigatorTree: Array<NavigatorTree>,
```

Comment:
In most places where you use a tree, the type is an array of trees. Is that really necessary? It could be a single tree with a root.
If it is really necessary, let's decide if we use singular or plural in the name, because e.g. getNavigatorRowsForTree is singular, but getNavigatorTrees is plural, although they use the same type.

Reply:
I followed the inconsistency of the `elementPathTree: ElementPathTrees` :D